File: pkg/rnib/rnib.go
package rnib

import (
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"strings"

	"gerrit.o-ran-sc.org/r/ric-plt/xapp-frame/pkg/xapp"
	"github.com/muriloAvlis/usap-5g/pkg/utils"
)

// Get eNBs list
func geteNBList() ([]*xapp.RNIBNbIdentity, error) {
	eNBs, err := xapp.Rnib.GetListEnbIds()
	if err != nil {
		xapp.Logger.Error("Unable to get eNodeB list: %s", err.Error())
		return nil, err
	}

	var eNB_names []string
	for _, eNB := range eNBs {
		eNB_names = append(eNB_names, eNB.GetInventoryName())
	}

	xapp.Logger.Info("List of connected eNodeBs: [%s]", strings.Join(eNB_names, ", "))

	return eNBs, nil
}

// Get gNBs list
func getgNBList() ([]*xapp.RNIBNbIdentity, error) {
	gNBs, err := xapp.Rnib.GetListGnbIds()
	if err != nil {
		xapp.Logger.Error("Unable to get gNodeB list: %s", err.Error())
		return nil, err
	}

	var gNB_names []string
	for _, gNB := range gNBs {
		gNB_names = append(gNB_names, gNB.GetInventoryName())
	}

	xapp.Logger.Info("List of connected gNodeBs: [%s]", strings.Join(gNB_names, ", "))

	return gNBs, nil
}

// Get gNB and eNB list connected to RIC
func GetNbList() []*xapp.RNIBNbIdentity {
	var nodeBs []*xapp.RNIBNbIdentity

	if eNBs, err := geteNBList(); err == nil {
		nodeBs = append(nodeBs, eNBs...)
	}

	if gNBs, err := getgNBList(); err == nil {
		nodeBs = append(nodeBs, gNBs...)
	}

	return nodeBs
}

func GetRanFuncDefiniton(inventoryName string, ranFuncId int64) (string, error) {
	uri := "http://" + os.Getenv("E2MGR_HTTP_SERVICE_HOST") + ":" + os.Getenv("E2MGR_HTTP_SERVICE_PORT") + "/v1/nodeb/" + inventoryName

	response, err := http.Get(uri)
	if err != nil {
		return "", fmt.Errorf("failed to get E2 node informations from E2MGR: %s", err.Error())
	}
	defer response.Body.Close()

	// Decode E2 node response
	var e2Response E2mgrResponse
	err = json.NewDecoder(response.Body).Decode(&e2Response)
	if err != nil {
		return "", fmt.Errorf("failed to decode E2 node informations from E2MGR: %s", err.Error())
	}

	// check if E2 Node has KPM RAN function == 2
	rfIdx := -1
	for idx, rf := range e2Response.Gnb.RanFunctions {
		if rf.RanFunctionId == ranFuncId {
			rfIdx = idx
			xapp.Logger.Debug("E2 node %s has KPM RF index: %d", inventoryName, rfIdx)
		}
	}

	if rfIdx == -1 {
		return "", fmt.Errorf("e2 node %s does not have KPM RF, finalizing KPI extraction proccess", inventoryName)
	}

	return e2Response.Gnb.RanFunctions[rfIdx].RanFunctionDefinition, nil
}

func GetMeasNameList(ranFuncDefinition map[string]interface{}, reportStyleType int) []string {
	var measNameList []string

	reportStyles := ranFuncDefinition["ric-ReportStyle-List"].([]interface{})

	for _, reportStyle := range reportStyles {
		reportStyleMap := reportStyle.(map[string]interface{})

		styleTypeInt := utils.FloatInterfaceToInt(reportStyleMap["ric-ReportStyle-Type"])

		if styleTypeInt == reportStyleType {
			measList := reportStyleMap["measInfo-Action-List"].([]interface{})
			for _, meas := range measList {
				measMap := meas.(map[string]interface{})
				measNameList = append(measNameList, measMap["measName"].(string))
			}
		}
	}

	return measNameList
}

func GetDuId(inventoryName string) (string, error) {
	uri := "http://" + os.Getenv("E2MGR_HTTP_SERVICE_HOST") + ":" + os.Getenv("E2MGR_HTTP_SERVICE_PORT") + "/v1/nodeb/" + inventoryName

	response, err := http.Get(uri)
	if err != nil {
		return "", fmt.Errorf("failed to get E2 node informations from E2MGR: %s", err.Error())
	}
	defer response.Body.Close()

	// Decode E2 node response
	var e2Response E2mgrResponse
	err = json.NewDecoder(response.Body).Decode(&e2Response)
	if err != nil {
		return "", fmt.Errorf("failed to decode E2 node informations from E2MGR: %s", err.Error())
	}

	// check if E2 Node has KPM RAN function == 2
	if e2Response.DuId == "" {
		return "", fmt.Errorf("e2 node %s has no DU ID", inventoryName)
	}

	return e2Response.DuId, nil
}

// Função para inverter uma string
func ReverseString(input string) string {
	runes := []rune(input)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

// Decode PLMN and return MCC and MNC
func DecodePLMN(plmn string) string {
	var mcc, mnc string

	if strings.Contains(plmn, "F") {
		// ABFCDE
		// BACF -> BAC (MCC)
		mcc = ReverseString(plmn[0:2]) + strings.ReplaceAll(ReverseString(plmn[2:4]), "F", "")

		// Invert 2 last digits (MNC)
		mnc = ReverseString(plmn[4:6])
	} else {
		// ABCDEF
		// ABCD -> BAD (MCC)
		mcc = ReverseString(plmn[0:2]) + ReverseString(string(plmn[3]))

		// FE + D (MNC)
		mnc = ReverseString(plmn[4:6]) + string(ReverseString(string(plmn[2]))[0])
	}

	return mcc + mnc // plmn
}

File: xapp/usap/rnib/rnib.py
import requests
import os

from usap.e2sm.utils import decode_plmn
from usap.config.config import Config


class Rnib(object):
    def __init__(self):
        self.logger = Config.get_logger()

        self.E2_MGR_HOST = os.getenv('E2MGR_HTTP_SERVICE_HOST')
        self.E2_MGR_PORT = os.getenv('E2MGR_HTTP_SERVICE_PORT')
        self.E2_MGR_URI = 'http://' + self.E2_MGR_HOST + \
            ':' + self.E2_MGR_PORT + '/v1/nodeb/'

    def get_nbs_list(self) -> list:
        nbList = []

        url = self.E2_MGR_URI + 'states'
        response = requests.get(url)

        if response.status_code == 200:
            self.logger.info('Get E2 Nodes from E2 Manager with success!')
            data = response.json()
            nbList.extend(data)

            for nb in nbList:
                # Convert to bool
                nb['connectionStatus'] = nb['connectionStatus'] == 'CONNECTED'

                # Change PLMN to decoded format
                mcc, mnc = decode_plmn(nb['globalNbId'].get('plmnId'))
                nb['globalNbId']['plmnId'] = mcc + mnc
        else:
            self.logger.warning(f'''Unable to get E2 nodes [{
                response.status_code}]''')

        return nbList

    def get_ran_func_def_by_invetoryName(self, inventory_name, ran_func_id):
        url = self.E2_MGR_URI + inventory_name
        response = requests.get(url)

        if response.status_code == 200:
            self.logger.info(
                'Get RAN Function Definition from E2 Manager with success!')
            data = response.json()
            ran_functions = data['gnb']['ranFunctions']

            for rf in ran_functions:
                if rf['ranFunctionId'] == ran_func_id:
                    return rf['ranFunctionDefinition']

        self.logger.warning(f'''Unable to get RAN Function Definition to E2 Node {inventory_name} [{
            response.status_code}]''')

        return None

    def get_metric_names_by_report_style(self, ran_func_def: dict, report_style_type: int):
        return [
            metric['measName']
            for report_style in ran_func_def.get('ric-ReportStyle-List', [])
            if report_style.get('ric-ReportStyle-Type') == report_style_type
            for metric in report_style.get('measInfo-Action-List', [])
        ]


# tests
# if __name__ == '__main__':
#     rnib = Rnib()
#     kpm_packer = e2sm_kpm_packer()
#     nbList = rnib.get_nbs_list()

#     for nb in nbList:
#         print(60 * '*')
#         inventory_name = nb['inventoryName']
#         print(f'RAN NAME: {inventory_name}')
#         rf_def = rnib.get_ran_func_def_by_invetoryName(inventory_name, 2)
#         rf_def_bytes = bytes.fromhex(rf_def)
#         rf_def_decoded = kpm_packer.decode_ran_function_definition(
#             rf_def_bytes)

#         metric_names = rnib.get_metric_names_by_report_style(rf_def_decoded, 4)
#         print(metric_names)
